### PR TITLE
Implement Biot-Savart variant

### PR DESCRIPTION
**Pull request by pazathoth**
A variant of the Biot-Savart law which yields the gauged vector potential toroidal modes directly should be implemented. The following additions are necessary:

1. A variant of `Biot_Savart_Fourier` in [`coil_tools`](../tree/main/src/magfie/coil_tools.f90) containing the actual computation,
2. `write_Anvac` and `read_Anvac`, also in `coil_tools`, to save/load this result,
3. an additional command line option for [`vacfield.x`](../tree/main/src/magfie/vacfield.f90) to call these subroutines,
4. a variant of `vector_potential_single_mode` in [`bdivfree`](../tree/main/src/magfie/bdivfree.f90) which splines the result.

@jonatanschatzlmayr already checked my derivation and started the implementation of point 1. Regarding point 2, I suggest a NetCDF file with `coil_index`, `ntor`, `R` and `Z` as dimensions for arrays `An_R` and `An_Z`; named coil groups can be added at a later point. For point 3, an option `gauged` (suggestions welcome) can be handled analogously to the option `Fourier`. Concerning point 4, my version does not use a limiting convex polygon before splining, which might need to be added as an option.

### CODE REVIEW

File: src/magfie/bdivfree.f90 [hunk 1]
```diff
@@ -181,6 +181,51 @@ subroutine vector_potentials(nr_in,np_in,nz_in,ntor_in,      &
 102 format(1000e15.7)
 end subroutine vector_potentials
 
+
+! Spline a single toroidal Fourier mode of the RMP coil field, as needed for MEPHIT.
+! Note the explicit-shape input arrays are not checked for shape.
+! Also, the truncation of data outside the limiting convex is not performed.
+subroutine vector_potential_single_mode(ntor_in, nR_in, nZ_in, Rmin_in, Rmax_in, Zmin_in, Zmax_in, Bn_R, Bn_Z)
```
Comment (krystophny) [inline src/magfie/bdivfree.f90:188]: If you spline this, why not call the routine `spline_vector_potential_single_mode`?

File: src/magfie/bdivfree.f90 [hunk 2]
```diff
@@ -181,6 +181,51 @@ subroutine vector_potentials(nr_in,np_in,nz_in,ntor_in,      &
...
+  use iso_fortran_env, only: dp => real64
+  use bdivfree_mod, only: nR, nZ, Rmin, Zmin, ntor, icp, ipoint, hR, hZ, Rpoi, Zpoi, AZnRe, AZnIm, ARnRe, ARnIm
```
Comment (krystophny) [inline src/magfie/bdivfree.f90:190]: Are all these stored variables needed later? Or can one read some quantities from the size of arrays to reduce the argument list and make everything less side-effecty?

File: src/magfie/bdivfree.f90 [hunk 3]
```diff
@@ -181,6 +181,51 @@ subroutine vector_potentials(nr_in,np_in,nz_in,ntor_in,      &
...
+  integer, intent(in) :: nR_in, nZ_in
+  real(dp), intent(in) :: Rmin_in, Rmax_in, Zmin_in, Zmax_in
+  complex(dp), intent(in), dimension(nR_in, nZ_in) :: Bn_R, Bn_Z
+  integer :: k, imi(nZ_in), ima(nZ_in), jmi(nR_in), jma(nR_in)
```
Comment (krystophny) [inline src/magfie/bdivfree.f90:194]: In know it's nice if the call to `s2dcut` fits one line, but still please use a bit more speaking variable names. I guess it means min and max of something, but what?

**Review by krystophny — CHANGES_REQUESTED** [review]

File: src/magfie/bdivfree.f90 (hunk 1 above)
Comment (pazathoth) [reply src/magfie/bdivfree.f90:188]: For consistency with the other routines in this file. It's a single-mode analogue of `vector_potentials`, while the `spline_*` routines are the ones to evaluate the splines. I don't know who else uses these functions.

File: src/magfie/bdivfree.f90 (hunk 2 above)
Comment (pazathoth) [reply src/magfie/bdivfree.f90:190]: They are used by the spline evaluation routines later and are saved just like in `vector_potentials`. To properly access array sizes, you would have to put these routines into a module (and give them proper names while you're at it). That's not the purpose of this pull request, but one of the distractions you would warn me about. But when we have time™ later, I can change it and you can test it.

File: src/magfie/bdivfree.f90 (hunk 3 above)
Comment (pazathoth) [reply src/magfie/bdivfree.f90:194]: These are the names used in `vector_potentials` and [`s2dcut`](../tree/main/src/magfie/spline5_RZ.f90), but I'm open to suggestions.

**Comment by pazathoth** [comment]
There seems to be a problem with OpenMP in `coil_tools`. When I link `magfie` against `OpenMP::OpenMP_Fortran`, I get the following error compiling locally:
```
/usr/bin/ld: __exchange_get_cancoord_mod_MOD_bctrvr_vartheta: TLS definition in ../src/libmagfie.so section .tbss mismatches non-TLS definition in ../liblibneo.so section .bss
/usr/bin/ld: ../src/libmagfie.so: error adding symbols: bad value
```
If I don't include it, it compiles locally, but at least in the Docker container, it doesn't link against OpenMP, causing build errors if OpenMP is not in `!$` comments.

File: src/magfie/coil_tools.f90 [hunk 4]
```diff
@@ -373,6 +374,270 @@ subroutine Biot_Savart_sum_coils(coils, Ic, &
...
+    call fftw_free(p_AnR)
+    call fftw_free(p_Anphi)
+    call fftw_free(p_AnZ)
+    call fftw_free(p_dAnphi_dR)
+    call fftw_free(p_dAnphi_dz)
+    !$ call fftw_cleanup_threads()
+    ! nullify pointers past this point
+  end subroutine Vector_Potential_Biot_Savart_Fourier
+
+  !dAXYZ(1,:) = dAXYZ_c - XYZ_if(1)/(dist_i*dist_f+sum(XYZ_i * XYZ_f))*(XYZ_i/dist_i + XYZ_f/dist_f) !(/Axx,Axy,Axz/)
```
Comment (krystophny) [inline src/magfie/coil_tools.f90:539]: why though?

File: src/magfie/vacfield.f90 [hunk 5]
```diff
@@ -41,22 +43,23 @@ program vacfield
...
-  Rmin = 75.0d0
-  Rmax = 267.0d0
-  Zmin = -154.0d0
-  Zmax = 150.4d0
-  nR = 150
-  nZ = 300
-  nphi = 512
-  nmax = 64
-  prefactor = 5 * current_si_to_cgs / C
+  ! Rmin = 75.0d0
```
Comment (krystophny) [inline src/magfie/vacfield.f90:53]: why though?

**Review by krystophny — APPROVED** [review]

**Pull request merged**